src/monotonic/monotonic.jl

```diff
@@ -0,0 +1,351 @@
+coefficients. Type is any concrete subtype of `MonotonicInterpolationType`.
+"""
+struct MonotonicInterpolation{T, TCoeffs, Tel, Type<:MonotonicInterpolationType,
+    K<:AbstractVector{<:Number}, AType <: AbstractArray{Tel,1}} <: AbstractInterpolation{T,1,DimSpec{Type}}
+
+    it::Type
+    knots::K
+    A::AType
+    m::Vector{TCoeffs}
+    c::Vector{TCoeffs}
+    d::Vector{TCoeffs}
+end
+
+
+size(A::MonotonicInterpolation) = size(A.knots)
+axes(A::MonotonicInterpolation) = axes(A.knots)
+
+function MonotonicInterpolation(::Type{TWeights}, it::IType, knots::K, A::AbstractArray{Tel,1},
+    m::Vector{TCoeffs}, c::Vector{TCoeffs}, d::Vector{TCoeffs}) where {TWeights, TCoeffs, Tel, IType<:MonotonicInterpolationType, K<:AbstractVector{<:Number}}
+
+    isconcretetype(IType) || error("The b-spline type must be a leaf type (was $IType)")
+    isconcretetype(TCoeffs) || warn("For performance reasons, consider using an array of a concrete type (eltype(A) == $(eltype(A)))")
+
+    check_monotonic(knots, A)
+
+    cZero = zero(TWeights)
+    if isempty(A)
+        T = Base.promote_op(*, typeof(cZero), eltype(A))
+    else
+        T = typeof(cZero * first(A))
+    end
+
+    MonotonicInterpolation{T, TCoeffs, Tel, IType, K, typeof(A)}(it, knots, A, m, c, d)
+end
+
+function interpolate(::Type{TWeights}, ::Type{TCoeffs}, knots::K,
+    A::AbstractArray{Tel,1}, it::IT) where {TWeights,TCoeffs,Tel,K<:AbstractVector{<:Number},IT<:MonotonicInterpolationType}
+
+    check_monotonic(knots, A)
+
+    # first we need to determine tangents (m)
+    n = length(knots)
+    m, Δ = calcTangents(TCoeffs, knots, A, it)
+    c = Vector{TCoeffs}(undef, n-1)
+    d = Vector{TCoeffs}(undef, n-1)
+    for k ∈ 1:n-1
+        if IT == LinearMonotonicInterpolation
+            c[k] = d[k] = zero(TCoeffs)
+        else
+            xdiff = knots[k+1] - knots[k]
+            c[k] = (3*Δ[k] - 2*m[k] - m[k+1]) / xdiff
+            d[k] = (m[k] + m[k+1] - 2*Δ[k]) / (xdiff * xdiff)
+        end
+    end
+
+    MonotonicInterpolation(TWeights, it, knots, A, m, c, d)
+end
+
+function interpolate(knots::AbstractVector{<:Number}, A::AbstractArray{Tel,1},
+    it::IT) where {Tel,N,IT<:MonotonicInterpolationType}
+
+    interpolate(tweight(A), tcoef(A), knots, A, it)
+end
+
+@inline function findKnot(itp::MonotonicInterpolation, x::Number)
+    x >= itp.knots[1] || error("Given number $x is outside of interpolated range [$(itp.knots[1]), $(itp.knots[end])]")
+    x <= itp.knots[end] || error("Given number $x is outside of interpolated range [$(itp.knots[1]), $(itp.knots[end])]")
+
+    k = 1
+    n = length(itp.knots)
+    while k < n-1 && x > itp.knots[k+1]
+        k += 1
+    end
+    return k
+end
+
+function (itp::MonotonicInterpolation)(x::Number)
+    k = findKnot(itp, x)
+    xdiff = x - itp.knots[k]
+    return itp.A[k] + itp.m[k]*xdiff + itp.c[k]*xdiff*xdiff + itp.d[k]*xdiff*xdiff*xdiff
+end
+
+function derivative(itp::MonotonicInterpolation, x::Number)
```

Comment:
Why is this function not named `gradient` (returning an array of length 1) or `gradient1` (returning a scalar)?

Reply:
I've used the convention from [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl/blob/master/src/derivative.jl) but `gradient1` seems like a more fitting choice. Should I write a new method for the `gradient` function as well (receiving and returning one element arrays)?

Reply:
The behavior of `gradient` should, in order to match how other interpolation schemes work, be

    function gradient(itp, x...)
        g = Vector{eltype(itp)}(undef, lenght(x))
        # fill g with values
        g
    end

I think (correct me if I'm wrong) that the monotonic implementation here only supports 1D interpolations, which means this would effectively be `gradient(itp, x) = [derivative(itp, x)]`. `gradient1` is just for convenience when the user knows it's a 1D interpolation object, but it can also avoid the array allocation: `gradient1(itp, x) = derivative(itp, x)`. In other words, if you rename `derivative` to `gradient1` and then add `gradient(itp, x) = [gradient1(itp, x)]` (with typed arguments) you'll be good.

As of #240 there's a fallback defined for `gradient!` that just calls `gradient`.

Reply:
But `gradient` should return an `SVector` rather than `Vector`---there's a ton of overhead for `Vector` but essentially none for `SVector`.

Reply:
I've renamed `derivative` to `gradient1` and introduced new functions: `gradient`, `hessian` and `hessian1`. `hessian` and `gradient` return `SVector`s.